# SexSpecificScaffolds_Otur.R code to quantify sex mapping bias across scaffolds
#   in Ornithodoros turicata female assembly 
# Takes female and male depth_by_scaffold tsv mapping files generated by 
#   mapping illumina gDNA reads from 3 individuals of each sex
# Tests for significant sex mapping bias across major scaffolds (chromosomes)


library(tidyverse)

ASSEMBLY = "GCF_037126465.1_ASM3712646v1"
MIN_SCAF_LENGTH = 10000000

ASM_FAI <- read.table("GCF_037126465.1_ASM3712646v1_genomic.fna.fai")
row.names(ASM_FAI) <- ASM_FAI$V1
v.SCAF_LENGTHS <- ASM_FAI$V2

# get names of scaffolds longer than MIN_SCAF_LENGTH
SCAFFOLDS <- ASM_FAI[ASM_FAI$V2 >= MIN_SCAF_LENGTH,]$V1

# Input Datasets
# depth_by_scaffold.tsv
IN.depth <- list(F1 = read.table("O_turi_F1.depth_by_scaffold.tsv", 
                                     header = F, row.names = 1, 
                                     col.names = c('scaffold', 'depth')),
                 F2 = read.table("O_turi_F2.depth_by_scaffold.tsv", 
                                      header = F, row.names = 1, 
                                      col.names = c('scaffold', 'depth')),
                 F3 = read.table("O_turi_F3.depth_by_scaffold.tsv", 
                                      header = F, row.names = 1, 
                                      col.names = c('scaffold', 'depth')),
                 M1 = read.table("O_turi_M1a.depth_by_scaffold.tsv", 
                                      header = F, row.names = 1, 
                                      col.names = c('scaffold', 'depth')),
                 M2 = read.table("O_turi_M2.depth_by_scaffold.tsv", 
                                      header = F, row.names = 1, 
                                      col.names = c('scaffold', 'depth')),
                 M3 = read.table("O_turi_M1b.depth_by_scaffold.tsv", 
                                      header = F, row.names = 1, 
                                      col.names = c('scaffold', 'depth')) )

# segregating sites
IN.n_seg <- list(F1 = read.csv("O_turi_F1.bam.nseg_by_scaffold.csv", header = F, 
                               col.names = c('n_seg', 'scaffold'), row.names = 2),
                 F2 = read.csv("O_turi_F2.bam.nseg_by_scaffold.csv", header = F, 
                               col.names = c('n_seg', 'scaffold'), row.names = 2),
                 F3 = read.csv("O_turi_F3.bam.nseg_by_scaffold.csv", header = F, 
                               col.names = c('n_seg', 'scaffold'), row.names = 2),
                 M1 = read.csv("O_turi_M1a.bam.nseg_by_scaffold.csv", header = F, 
                               col.names = c('n_seg', 'scaffold'), row.names = 2),
                 M2 = read.csv("O_turi_M2.bam.nseg_by_scaffold.csv", header = F, 
                               col.names = c('n_seg', 'scaffold'), row.names = 2),
                 M3 = read.csv("O_turi_M1b.bam.nseg_by_scaffold.csv", header = F, 
                               col.names = c('n_seg', 'scaffold'), row.names = 2) )


# Analysis

# Mean depth per library
sapply(IN.depth, function(x){ mean(x$depth)})
# Depths for mitochondria and chromosomes per library
lapply(IN.depth, function(x) head(x, n=11))
# Mean depth on assembled chromosomes per library
sapply(IN.depth, function(x){ mean(x$depth[2:(1+length(SCAFFOLDS))])})

# Read depth
 
# each libraries depths values (raw and mean-normalized) for long scaffolds 
l.ScafDepths <- lapply(IN.depth, function(x){
  data.frame(scaffold = SCAFFOLDS,
             raw_depth = x[SCAFFOLDS,],
             norm_depth = x[SCAFFOLDS,] / mean(x$depth) 
             # can also normalize only on the autosomes - 1 and 2 are mt and X
             , auto_norm_depth = x[SCAFFOLDS,] / mean(x$depth[3:nrow(x)])
            )
})

# combined into single data frame
df.ScafDepths <- bind_rows(l.ScafDepths, .id = "library")
# add column for sex ID
df.ScafDepths$sex <- rep("female", times = nrow(df.ScafDepths))
df.ScafDepths$sex[df.ScafDepths$library %in% c('M1','M2','M3')] <- "male"
# add column for rep number
df.ScafDepths$rep <- substr(df.ScafDepths$library, 2, 2)


# plot depths by scaffold/sex
ggplot(data = df.ScafDepths, aes(x=scaffold, y=auto_norm_depth, fill=sex)) + 
  geom_boxplot() + 
  geom_point(aes(shape=rep, group=sex),
             position = position_dodge(width = 0.75)) + 
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)
       # , legend.position = "inside", legend.position.inside = c(.08, .85)
        ) + 
  xlab(paste(ASSEMBLY, "scaffolds longer than", MIN_SCAF_LENGTH/1000000, "Mbp", 
             sep = " " )) +
  ylab("mean scaffold depth / mean library depth") 

# anova for each scaffold
l.ScafDepths_aov <- lapply(split(df.ScafDepths, df.ScafDepths$scaffold),
                            function(d){
                              aov( norm_depth ~ sex , data=d)
                            })
lapply(l.ScafDepths_aov, summary)

# Same tests but normalized based on all scaffolds except suspect-X
# plot depths by scaffold/sex
# ggplot(data = df.ScafDepths, aes(x=scaffold, y=auto_norm_depth, fill=sex)) +
#   geom_boxplot() +
#   geom_point(position = position_dodge(width = 0.75)) +
#   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

# anova for each scaffold, reads normalized with autosomes 
l.ScafDepths_aov_autonorm <- lapply(split(df.ScafDepths, df.ScafDepths$scaffold),
                           function(d){
                             aov( auto_norm_depth ~ sex , data=d)
                           })
lapply(l.ScafDepths_aov_autonorm, summary)
# adjusted p-values
sapply(l.ScafDepths_aov, function(x) summary(x)[[1]]["Pr(>F)"][[1]][1]) %>% 
  p.adjust(method="fdr")

# Segregating sites
# add split scaffold to the scaffold list, these the last 2 rows of n_seg dfs
SCAFFOLDS_wSplit <- c(SCAFFOLDS[1], 
                      rownames(IN.n_seg$F1)[(nrow(IN.n_seg$F1)-1):nrow(IN.n_seg$F1)],
                      SCAFFOLDS[2:length(SCAFFOLDS)])
# ordered lengths of scaffolds, including split NC_088201.1
LENGTHS_wSplit <- c(ASM_FAI[SCAFFOLDS, 2][1],
                    90000000, (ASM_FAI[SCAFFOLDS, 2][1]-90000000),
                    ASM_FAI[SCAFFOLDS, 2][2:length(SCAFFOLDS)])
# each libraries number and proportion segregating sites for long scaffolds 
l.ScafNSegs <- lapply(IN.n_seg, function(x){
  data.frame(scaffold = SCAFFOLDS_wSplit,
             n_seg = x[SCAFFOLDS_wSplit,],
             p_seg = x[SCAFFOLDS_wSplit,] / LENGTHS_wSplit )
})

# combined into single data frame
df.ScafNSegs <- bind_rows(l.ScafNSegs, .id = "library")
# add column for sex ID
df.ScafNSegs$sex <- rep("female", times = nrow(df.ScafNSegs))
df.ScafNSegs$sex[df.ScafNSegs$library %in% c('M1','M2','M3')] <- "male"
# add column for rep number
df.ScafNSegs$rep <- substr(df.ScafNSegs$library, 2, 2)

# range of percent segregating
range(df.ScafNSegs[df.ScafNSegs$scaffold %in% SCAFFOLDS,]$p_seg)

# variant density on chromosome 1 by sex, region
df.ScafNSegs %>% filter(grepl('NC_088201.1', scaffold)) %>% 
  group_by(scaffold, sex) %>% 
  summarise(var_per_kb = mean(p_seg)*1000)

# plot segregation rate by scaffold/sex
ggplot(data = df.ScafNSegs, aes(x=scaffold, y=p_seg, fill=sex)) + 
  geom_boxplot() + 
  geom_point(aes(shape=rep, group=sex),
             position = position_dodge(width = 0.75)) + 
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)
        #, legend.position = "inside", legend.position.inside = c(.08, .85)
        ) + 
  xlab(paste(ASSEMBLY, "scaffolds longer than", MIN_SCAF_LENGTH/1000000, "Mbp", 
             sep = " " )) +
  ylab("number segregating sites / scaffold length") 

# anova for each scaffold region (skip the unsplit contig so each region tested once)
df.ScafNSegs_NoUnsplit <- df.ScafNSegs[df.ScafNSegs$scaffold != "NC_088201.1",]
l.ScafNSegs_aov <- lapply(split(df.ScafNSegs_NoUnsplit, df.ScafNSegs_NoUnsplit$scaffold),
                           function(d){
                             aov( p_seg ~ sex , data=d)
                           })
lapply(l.ScafNSegs_aov, summary)
# adjusted p values
sapply(l.ScafNSegs_aov, function(x) summary(x)[[1]]["Pr(>F)"][[1]][1]) %>% 
  p.adjust(method="fdr")
